chainerx_cc/chainerx/native/reduce.h

```diff
@@ -11,19 +11,83 @@ namespace chainerx {
 namespace native {
 namespace reduce_detail {
 
+// The number of items those are processed by the statically expanded pairwise reduction routine.
+// Reduction performance is sensitive to this parameter. Increasing the parameter may improve performance,
+// but it can also severely degrade the performance due to register pressure.
+// Must be a power of 2.
+constexpr int64_t ExpandLen = 4;
+// The number of items those are reduced by the serial loop. The loop runs on the outside of the statically expanded
+// reduction. A large value makes overhead for pairwise reduction small. So increasing this parameter will improve
+// performance a little, in exchange for numerical precision.
+// Must be a power of 2.
+constexpr int64_t SerialLen = 8;
+
+template <typename In, typename ReductionImpl, int8_t InNdim, typename T, int64_t n>
+struct ExpandedPairwiseReduction {
+    static T run(const IndexableArray<const In, InNdim>& in, IndexIterator<InNdim>& it_in, ReductionImpl&& impl, int64_t& i_reduce) {
+        T accum = ExpandedPairwiseReduction<In, ReductionImpl, InNdim, T, n / 2>::run(in, it_in, impl, i_reduce);
+        impl.Reduce(ExpandedPairwiseReduction<In, ReductionImpl, InNdim, T, n / 2>::run(in, it_in, impl, i_reduce), accum);
+        return accum;
+    }
+};
+
+template <typename In, typename ReductionImpl, int8_t InNdim, typename T>
+struct ExpandedPairwiseReduction<In, ReductionImpl, InNdim, T, 1> {
+    static T run(const IndexableArray<const In, InNdim>& in, IndexIterator<InNdim>& it_in, ReductionImpl&& impl, int64_t& i_reduce) {
+        T accum = impl.MapIn(native_internal::StorageToDataType<const In>(in[it_in]), i_reduce);
+        ++it_in, ++i_reduce;
+        return accum;
+    }
+};
+
+constexpr int log2(int64_t v) { return v == 1 ? 0 : log2(v >> 1) + 1; }
+
+template <typename In, typename ReductionImpl, int8_t InNdim, typename T>
+T PairwiseReduction(const IndexableArray<const In, InNdim>& in, IndexIterator<InNdim>& it_in, ReductionImpl&& impl, int64_t reduce_len) {
+    int64_t i_reduce = 0;
+    T accum = impl.Identity();
+    T tree_accum[63 - log2(ExpandLen * SerialLen)];
```

Comment:
Could you fix for cpplint?
https://travis-ci.org/chainer/chainer/jobs/529116502#L462

Reply:
Sure. I'll introduce a temporary variable. NOLINT is inappropriate, because non-constexpr expression may slip into here in future.

Reply:
Now fixed.